<a href="https://colab.research.google.com/github/DevHyunseo/PoscoAcademy_Bigdata/blob/main/%ED%94%84%EB%A1%9C%EC%A0%9D%ED%8A%B83_%EC%A2%85%ED%95%A9%EC%8B%A4%EC%8A%B5(%EB%8C%80%ED%98%95%EB%A7%88%ED%8A%B8)_%EB%8D%B0%EC%9D%B4%ED%84%B0_%ED%99%95%EC%9D%B8_%EB%B0%8F_%EC%A0%84%EC%B2%98%EB%A6%AC.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib
import matplotlib.pyplot as plt
import matplotlib.pyplot as plt
from matplotlib import font_manager, rc
from datetime import datetime
import warnings

warnings.filterwarnings('ignore')

sns.set_palette('Pastel1')
plt.rcParams['font.family'] = 'NanumGothic'
matplotlib.rc("font",family = "NanumGothic")

# 1. 데이터 불러오기

### (1) 고객 정보 데이터

- 고객 정보 데이터(Row : 11,555 / Column : 8) | 파일명 : member_data.csv

| 변수명             | 타입   | 상세내용                                                 |
| ------------------ | ------ | -------------------------------------------------------- |
| 회원번호           | ID     | 회원 고유 식별 번호                                      |
| 조합원상태         | 범주형 | 회원 가입 상태(탈퇴여부)                                 |
| 주소               | 범주형 | 회원 주소지                                              |
| 성별               | 범주형 | 회원성별                                                 |
| 연령               | 연속형 | 회원 연령                                                |
| 배송서비스신청여부 | 범주형 | 매장에서 물품 구매 시 집까지 배송서비스 신청 여부        |
| 모바일알람신청여부 | 범주형 | 프로모션 및 할인 관련, 마케팅 관련 메시지 알람 수신 여부 |
| Gold_member        | 범주형 | VIP 고객 여부                                            |


### (2) 판매 정보 데이터

- (Row : 123,191 / Column : 12) | 파일명 product_data_001.csv

- (Row : 55,108 / Column : 12) | 파일명 product_data_002.csv

- (Row : 63,776 / Column : 12) | 파일명 product_data_003.csv

- (Row : 52,570 / Column : 12) | 파일명 product_data_004.csv

| 변수명           | 타입     | 상세내용                                    |
| ---------------- | -------- | ------------------------------------------- |
| 공급일자         | datetime | 상품 판매 일자                              |
| 회원번호         | ID       | 회원 고유 식별 ID                           |
| 조합원상태       | 범주형   | 회원 가입 상태(탈퇴여부)                    |
| 물품대분류       | 범주형   | 구매 제품의 물품 대분류                     |
| 물품중분류       | 범주형   | 구매 제품의 물품 중분류                     |
| 물품소분류       | 범주형   | 구매 제품의 물품 소분류                     |
| 물품명           | 범주형   | 구매 제품의 물품명                          |
| 구매수량         | 연속형   | 고객이 구매한 물품 수량                     |
| 구매금액         | 연속형   | 고객이 구매한 물품 금액( 원단위)            |
| 구매매장         | 범주형   | 구매 매장                                   |
| 반품일자 | datetime | 고객이 품목을 반품한 경우, 반품 날짜 (환불) |
| 구매시각         | time     | 구매시각 (바코드에 상품이 찍힌 시간)        |

# 2. 데이터 확인하기

### (1) 고객 정보 데이터 확인
- 고객 정보 데이터 로드
- 데이터 shape, info, describe 확인

In [ ]:
df = pd.read_csv("/home/piai/바탕화면/빅데이터주간/A4_대형마트/member_data.csv")
df

회원번호 조합원상태        주소 성별    연령 배송서비스신청여부 모바일알람여부 Gold_member
0      1000012800  정상회원  수지구-풍덕천동  여  58.0       미신청       .      normal
1      1000103744  정상회원   처인구-고림동  여  34.0       미신청       .      normal
2      1000122304  정상회원     광주-기타  여  38.0       미신청       .      normal
3      1000163136  정상회원   수지구-동천동  여  73.0       미신청       .      normal
4      1000281920  정상회원  수지구-풍덕천동  여  42.0        신청       .      normal
...           ...   ...       ... ..   ...       ...     ...         ...
11550   999682432    탈퇴   수지구-동천동  여  47.0       미신청       .      normal
11551   999721408  정상회원   수지구-동천동  여  43.0       미신청       .      normal
11552   999738112  정상회원   수지구-성복동  여  67.0       미신청       .      normal
11553   999754816  정상회원   수지구-상현동  여  44.0        신청       .      normal
11554    99995712  정상회원   수지구-상현동  여  66.0       미신청       .      normal

[11555 rows x 8 columns]

In [ ]:
df.shape

(11555, 8)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11555 entries, 0 to 11554
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   회원번호         11555 non-null  int64  
 1   조합원상태        11555 non-null  object 
 2   주소           11555 non-null  object 
 3   성별           11553 non-null  object 
 4   연령           11553 non-null  float64
 5   배송서비스신청여부    11555 non-null  object 
 6   모바일알람여부      11555 non-null  object 
 7   Gold_member  11555 non-null  object 
dtypes: float64(1), int64(1), object(6)
memory usage: 722.3+ KB


In [ ]:
df.describe(include='all')

회원번호  조합원상태        주소     성별            연령 배송서비스신청여부 모바일알람여부  \
count   1.155500e+04  11555     11555  11553  11553.000000     11555   11555   
unique           NaN      4        95      2           NaN         2       2   
top              NaN   정상회원  수지구-풍덕천동      여           NaN       미신청       .   
freq             NaN  11413      3188  10885           NaN     10512   10653   
mean    8.157145e+08    NaN       NaN    NaN     49.704233       NaN     NaN   
std     4.409753e+08    NaN       NaN    NaN     11.563223       NaN     NaN   
min     1.861382e+07    NaN       NaN    NaN     17.000000       NaN     NaN   
25%     4.567941e+08    NaN       NaN    NaN     41.000000       NaN     NaN   
50%     7.718696e+08    NaN       NaN    NaN     48.000000       NaN     NaN   
75%     1.170548e+09    NaN       NaN    NaN     56.000000       NaN     NaN   
max     1.670621e+09    NaN       NaN    NaN    108.000000       NaN     NaN   

       Gold_member  
count        11555  
unique           2  
top         normal  
freq         11350  
mean           NaN  
std            NaN  
min            NaN  
25%            NaN  
50%            NaN  
75%            NaN  
max            NaN

## (2) 판매 정보 데이터 확인
- 판매 정보 데이터 로드
- 데이터 shape, info, describe 확인
- A, B, C, D 매장에 대한 정보 확인 후 합치기(merge)

In [ ]:
df1 = pd.read_csv("/home/piai/바탕화면/빅데이터주간/A4_대형마트/product_data_001.csv", encoding = 'euc-kr')
df2 = pd.read_csv("/home/piai/바탕화면/빅데이터주간/A4_대형마트/product_data_002.csv", encoding = 'euc-kr')
df3 = pd.read_csv("/home/piai/바탕화면/빅데이터주간/A4_대형마트/product_data_003.csv", encoding = 'euc-kr')
df4 = pd.read_csv("/home/piai/바탕화면/빅데이터주간/A4_대형마트/product_data_004.csv", encoding = 'euc-kr')

In [ ]:
df1.head()

Unnamed: 0.1  Unnamed: 0        공급일자        회원번호 조합원상태  물품대분류 물품중분류  \
0             0           0  2023-01-06   272369856  정상회원     과실    과일   
1             1           1  2023-01-06  1476143616  정상회원     간식     빵   
2             2           2  2023-01-06   743192512  정상회원     채소  양념채소   
3             3           3  2023-01-06   176912064  정상회원    축산물   유제품   
4             4           4  2023-01-06    87592064  정상회원  양념/가루    가루   

      물품소분류                  물품명  구매수량   구매금액 구매매장   구매시각 반품일자  
0        사과          사과/유(1.5kg)   1.0  22207  매장C  10:04  NaN  
1  식빵, 식사대용      아침빵(240g/8개)/매장   1.0   4403  매장C  10:09  NaN  
2        생강             생강(300g)   1.0   7083  매장C  10:12  NaN  
3    우유/산양유  유기농저지방우유(종이팩/900ml)   1.0   7849  매장C  10:23  NaN  
4      양념가루        카레(매운맛: 100g)   3.0  15507  매장C  10:29  NaN

In [ ]:
df1.shape

(123191, 14)

In [ ]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 123191 entries, 0 to 123190
Data columns (total 14 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   Unnamed: 0.1  123191 non-null  int64  
 1   Unnamed: 0    123191 non-null  int64  
 2   공급일자          123191 non-null  object 
 3   회원번호          123191 non-null  int64  
 4   조합원상태         123191 non-null  object 
 5   물품대분류         123191 non-null  object 
 6   물품중분류         123191 non-null  object 
 7   물품소분류         123191 non-null  object 
 8   물품명           123191 non-null  object 
 9   구매수량          123191 non-null  float64
 10  구매금액          123191 non-null  int64  
 11  구매매장          123191 non-null  object 
 12  구매시각          123191 non-null  object 
 13  반품일자          1789 non-null    object 
dtypes: float64(1), int64(4), object(9)
memory usage: 13.2+ MB


In [ ]:
df1.describe(include='all')

Unnamed: 0.1     Unnamed: 0        공급일자          회원번호   조합원상태  \
count   123191.000000  123191.000000      123191  1.231910e+05  123191   
unique            NaN            NaN         295           NaN       4   
top               NaN            NaN  2023-08-17           NaN    정상회원   
freq              NaN            NaN        1240           NaN  121693   
mean     61595.000000   61595.000000         NaN  7.184267e+08     NaN   
std      35562.322843   35562.322843         NaN  4.357415e+08     NaN   
min          0.000000       0.000000         NaN  1.861382e+07     NaN   
25%      30797.500000   30797.500000         NaN  3.424877e+08     NaN   
50%      61595.000000   61595.000000         NaN  7.177839e+08     NaN   
75%      92392.500000   92392.500000         NaN  1.052064e+09     NaN   
max     123190.000000  123190.000000         NaN  1.670606e+09     NaN   

         물품대분류   물품중분류   물품소분류          물품명           구매수량          구매금액  \
count   123191  123191  123191       123191  123191.000000  1.231910e+05   
unique      15      63     375         2038            NaN           NaN   
top         채소   두부/유부      두부  두부(1모:420g)            NaN           NaN   
freq     32352    9672    8753         4725            NaN           NaN   
mean       NaN     NaN     NaN          NaN       1.136326  1.079656e+04   
std        NaN     NaN     NaN          NaN       0.690563  1.672424e+04   
min        NaN     NaN     NaN          NaN     -30.000000 -6.662160e+05   
25%        NaN     NaN     NaN          NaN       1.000000  4.020000e+03   
50%        NaN     NaN     NaN          NaN       1.000000  7.275000e+03   
75%        NaN     NaN     NaN          NaN       1.000000  1.263500e+04   
max        NaN     NaN     NaN          NaN      30.000000  1.837838e+06   

          구매매장    구매시각        반품일자  
count   123191  123191        1789  
unique       1     647         286  
top        매장C   10:10  2023-08-18  
freq    123191     463          28  
mean       NaN     NaN         NaN  
std        NaN     NaN         NaN  
min        NaN     NaN         NaN  
25%        NaN     NaN         NaN  
50%        NaN     NaN         NaN  
75%        NaN     NaN         NaN  
max        NaN     NaN         NaN

In [ ]:
df2.head()

Unnamed: 0.1  Unnamed: 0        공급일자       회원번호 조합원상태 물품대분류   물품중분류 물품소분류  \
0             0           0  2023-01-06  133834304  정상회원    잡곡  콩/화본/깨    화본   
1             1           1  2023-01-06  749556736  정상회원    잡곡  콩/화본/깨    토종   
2             2           2  2023-01-06  787320768  정상회원    채소  잎/줄기채소   시금치   
3             3           3  2023-01-06  175843008  정상회원    채소    뿌리채소    당근   
4             4           4  2023-01-06  165261952  정상회원    간식     밀가공    만두   

           물품명  구매수량   구매금액 구매매장   구매시각 반품일자  
0   귀리(무/500g)   1.0   8232  매장D  10:32  NaN  
1  선비콩/유(500g)   1.0  17038  매장D  10:44  NaN  
2    시금치(300g)   1.0   4403  매장D  11:10  NaN  
3     당근(500g)   2.0   5360  매장D  11:45  NaN  
4    군만두(500g)   1.0   8998  매장D  12:14  NaN

In [ ]:
df2.shape

(55108, 14)

In [ ]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 55108 entries, 0 to 55107
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Unnamed: 0.1  55108 non-null  int64  
 1   Unnamed: 0    55108 non-null  int64  
 2   공급일자          55108 non-null  object 
 3   회원번호          55108 non-null  int64  
 4   조합원상태         55108 non-null  object 
 5   물품대분류         55108 non-null  object 
 6   물품중분류         55108 non-null  object 
 7   물품소분류         55108 non-null  object 
 8   물품명           55108 non-null  object 
 9   구매수량          55108 non-null  float64
 10  구매금액          55108 non-null  int64  
 11  구매매장          55108 non-null  object 
 12  구매시각          55108 non-null  object 
 13  반품일자          1284 non-null   object 
dtypes: float64(1), int64(4), object(9)
memory usage: 5.9+ MB


In [ ]:
df2.describe(include='all')

Unnamed: 0.1    Unnamed: 0        공급일자          회원번호  조합원상태  물품대분류  \
count   55108.000000  55108.000000       55108  5.510800e+04  55108  55108   
unique           NaN           NaN         295           NaN      3     16   
top              NaN           NaN  2023-08-27           NaN   정상회원     채소   
freq             NaN           NaN        1071           NaN  54443  15004   
mean    27553.500000  27553.500000         NaN  6.873551e+08    NaN    NaN   
std     15908.453654  15908.453654         NaN  4.031067e+08    NaN    NaN   
min         0.000000      0.000000         NaN  1.876416e+07    NaN    NaN   
25%     13776.750000  13776.750000         NaN  3.908829e+08    NaN    NaN   
50%     27553.500000  27553.500000         NaN  6.577200e+08    NaN    NaN   
75%     41330.250000  41330.250000         NaN  9.354314e+08    NaN    NaN   
max     55107.000000  55107.000000         NaN  1.670400e+09    NaN    NaN   

         물품중분류  물품소분류          물품명          구매수량          구매금액   구매매장   구매시각  \
count    55108  55108        55108  55108.000000  5.510800e+04  55108  55108   
unique      61    368         1928           NaN           NaN      1    634   
top     잎/줄기채소     두부  두부(1모:420g)           NaN           NaN    매장D  10:23   
freq      3875   3181         1936           NaN           NaN  55108    206   
mean       NaN    NaN          NaN      1.134665  1.137432e+04    NaN    NaN   
std        NaN    NaN          NaN      0.861883  1.800188e+04    NaN    NaN   
min        NaN    NaN          NaN    -30.000000 -2.019710e+05    NaN    NaN   
25%        NaN    NaN          NaN      1.000000  4.135000e+03    NaN    NaN   
50%        NaN    NaN          NaN      1.000000  7.658000e+03    NaN    NaN   
75%        NaN    NaN          NaN      1.000000  1.378400e+04    NaN    NaN   
max        NaN    NaN          NaN     40.000000  1.688514e+06    NaN    NaN   

              반품일자  
count         1284  
unique         267  
top     2023-08-28  
freq            56  
mean           NaN  
std            NaN  
min            NaN  
25%            NaN  
50%            NaN  
75%            NaN  
max            NaN

In [ ]:
df3.head()

Unnamed: 0.1  Unnamed: 0        공급일자        회원번호 조합원상태  물품대분류   물품중분류  \
0             0           0  2023-01-06   345017408  정상회원  양념/가루       장   
1             1           1  2023-01-06   395094144  정상회원     간식       빵   
2             2           2  2023-01-06   735399168  정상회원    축산물       알   
3             3           3  2023-01-06  1226079168  정상회원     반찬    어묵/묵   
4             4           4  2023-01-06   748032960  정상회원     채소  잎/줄기채소   

      물품소분류                  물품명  구매수량  구매금액 구매매장   구매시각 반품일자  
0     장류/기타            청국장(150g)   1.0  4786  매장B  10:20  NaN  
1  식빵, 식사대용      아침빵(240g/8개)/매장   1.0  4403  매장B  10:38  NaN  
2       유정란  동물복지유정란/중란(10알/국내산)   1.0  5552  매장B  10:59  NaN  
3        어묵          명태참어묵(300g)   1.0  7275  매장B  11:19  NaN  
4       시금치            시금치(300g)   1.0  4403  매장B  11:30  NaN

In [ ]:
df3.shape

(63776, 14)

In [ ]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 63776 entries, 0 to 63775
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Unnamed: 0.1  63776 non-null  int64  
 1   Unnamed: 0    63776 non-null  int64  
 2   공급일자          63776 non-null  object 
 3   회원번호          63776 non-null  int64  
 4   조합원상태         63776 non-null  object 
 5   물품대분류         63776 non-null  object 
 6   물품중분류         63776 non-null  object 
 7   물품소분류         63776 non-null  object 
 8   물품명           63776 non-null  object 
 9   구매수량          63776 non-null  float64
 10  구매금액          63776 non-null  int64  
 11  구매매장          63776 non-null  object 
 12  구매시각          63776 non-null  object 
 13  반품일자          860 non-null    object 
dtypes: float64(1), int64(4), object(9)
memory usage: 6.8+ MB


In [ ]:
df3.describe(include='all')

Unnamed: 0.1    Unnamed: 0        공급일자          회원번호  조합원상태  물품대분류  \
count   63776.000000  63776.000000       63776  6.377600e+04  63776  63776   
unique           NaN           NaN         295           NaN      3     15   
top              NaN           NaN  2023-07-20           NaN   정상회원     채소   
freq             NaN           NaN         793           NaN  63024  15842   
mean    31887.500000  31887.500000         NaN  8.107350e+08    NaN    NaN   
std     18410.689721  18410.689721         NaN  4.321675e+08    NaN    NaN   
min         0.000000      0.000000         NaN  1.939149e+07    NaN    NaN   
25%     15943.750000  15943.750000         NaN  4.551153e+08    NaN    NaN   
50%     31887.500000  31887.500000         NaN  8.325422e+08    NaN    NaN   
75%     47831.250000  47831.250000         NaN  1.120866e+09    NaN    NaN   
max     63775.000000  63775.000000         NaN  1.670621e+09    NaN    NaN   

        물품중분류  물품소분류          물품명          구매수량           구매금액   구매매장   구매시각  \
count   63776  63776        63776  63776.000000   63776.000000  63776  63776   
unique     61    367         1958           NaN            NaN      1    653   
top     두부/유부     두부  두부(1모:420g)           NaN            NaN    매장B  10:11   
freq     4506   3973         2415           NaN            NaN  63776    287   
mean      NaN    NaN          NaN      1.157116   11207.457351    NaN    NaN   
std       NaN    NaN          NaN      0.728093   16835.822082    NaN    NaN   
min       NaN    NaN          NaN    -10.000000 -166554.000000    NaN    NaN   
25%       NaN    NaN          NaN      1.000000    4020.000000    NaN    NaN   
50%       NaN    NaN          NaN      1.000000    7466.000000    NaN    NaN   
75%       NaN    NaN          NaN      1.000000   13401.000000    NaN    NaN   
max       NaN    NaN          NaN     25.000000  804054.000000    NaN    NaN   

              반품일자  
count          860  
unique         240  
top     2023-07-21  
freq            24  
mean           NaN  
std            NaN  
min            NaN  
25%            NaN  
50%            NaN  
75%            NaN  
max            NaN

In [ ]:
df4.head()

Unnamed: 0.1  Unnamed: 0        공급일자       회원번호 조합원상태 물품대분류  물품중분류   물품소분류  \
0             0           0  2023-01-06  220954944  정상회원    반찬  두부/유부      두부   
1             1           1  2023-01-06   89023040  정상회원    채소   열매채소  단호박/손질   
2             2           2  2023-01-06  354941440  정상회원   축산물      알     유정란   
3             3           3  2023-01-06  786634048    탈퇴    반찬   어묵/묵       묵   
4             4           4  2023-01-06  438543104  정상회원   축산물      알     유정란   

                물품명  구매수량  구매금액 구매매장   구매시각 반품일자  
0       두부(1모:420g)   1.0  3829  매장A  10:32  NaN  
1    단호박(깍뚝썰기/300g)   1.0  4977  매장A  10:41  NaN  
2  유정란/매장용(10알/국내산)   1.0  7083  매장A  11:00  NaN  
3     도토리묵(420g:1모)   1.0  7083  매장A  11:28  NaN  
4  유정란/매장용(10알/국내산)   1.0  7083  매장A  11:59  NaN

In [ ]:
df4.shape

(52570, 14)

In [ ]:
df4.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52570 entries, 0 to 52569
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Unnamed: 0.1  52570 non-null  int64  
 1   Unnamed: 0    52570 non-null  int64  
 2   공급일자          52570 non-null  object 
 3   회원번호          52570 non-null  int64  
 4   조합원상태         52570 non-null  object 
 5   물품대분류         52570 non-null  object 
 6   물품중분류         52570 non-null  object 
 7   물품소분류         52570 non-null  object 
 8   물품명           52570 non-null  object 
 9   구매수량          52570 non-null  float64
 10  구매금액          52570 non-null  int64  
 11  구매매장          52570 non-null  object 
 12  구매시각          52570 non-null  object 
 13  반품일자          1027 non-null   object 
dtypes: float64(1), int64(4), object(9)
memory usage: 5.6+ MB


In [ ]:
df4.describe(include='all')

Unnamed: 0.1    Unnamed: 0        공급일자          회원번호  조합원상태  물품대분류  \
count   52570.000000  52570.000000       52570  5.257000e+04  52570  52570   
unique           NaN           NaN         295           NaN      2     15   
top              NaN           NaN  2023-07-27           NaN   정상회원     채소   
freq             NaN           NaN         878           NaN  52211  12767   
mean    26284.500000  26284.500000         NaN  7.574913e+08    NaN    NaN   
std     15175.796163  15175.796163         NaN  4.355470e+08    NaN    NaN   
min         0.000000      0.000000         NaN  1.876416e+07    NaN    NaN   
25%     13142.250000  13142.250000         NaN  3.832018e+08    NaN    NaN   
50%     26284.500000  26284.500000         NaN  7.479420e+08    NaN    NaN   
75%     39426.750000  39426.750000         NaN  1.059544e+09    NaN    NaN   
max     52569.000000  52569.000000         NaN  1.670400e+09    NaN    NaN   

         물품중분류  물품소분류          물품명          구매수량          구매금액   구매매장   구매시각  \
count    52570  52570        52570  52570.000000  5.257000e+04  52570  52570   
unique      61    368         1985           NaN           NaN      1    645   
top     잎/줄기채소     두부  두부(1모:420g)           NaN           NaN    매장A  10:32   
freq      3272   2886         1678           NaN           NaN  52570    154   
mean       NaN    NaN          NaN      1.145004  1.110015e+04    NaN    NaN   
std        NaN    NaN          NaN      0.855142  2.053302e+04    NaN    NaN   
min        NaN    NaN          NaN    -50.000000 -5.456080e+05    NaN    NaN   
25%        NaN    NaN          NaN      1.000000  4.212000e+03    NaN    NaN   
50%        NaN    NaN          NaN      1.000000  7.466000e+03    NaN    NaN   
75%        NaN    NaN          NaN      1.000000  1.340100e+04    NaN    NaN   
max        NaN    NaN          NaN     50.000000  2.890766e+06    NaN    NaN   

              반품일자  
count         1027  
unique         250  
top     2023-07-28  
freq            37  
mean           NaN  
std            NaN  
min            NaN  
25%            NaN  
50%            NaN  
75%            NaN  
max            NaN

In [ ]:
# 매장 A, B, C, D 합친 dfm 데이터 생성
dfm = pd.concat([df4, df3, df1, df2], ignore_index=True)
dfm.head()

Unnamed: 0.1  Unnamed: 0        공급일자       회원번호 조합원상태 물품대분류  물품중분류   물품소분류  \
0             0           0  2023-01-06  220954944  정상회원    반찬  두부/유부      두부   
1             1           1  2023-01-06   89023040  정상회원    채소   열매채소  단호박/손질   
2             2           2  2023-01-06  354941440  정상회원   축산물      알     유정란   
3             3           3  2023-01-06  786634048    탈퇴    반찬   어묵/묵       묵   
4             4           4  2023-01-06  438543104  정상회원   축산물      알     유정란   

                물품명  구매수량  구매금액 구매매장   구매시각 반품일자  
0       두부(1모:420g)   1.0  3829  매장A  10:32  NaN  
1    단호박(깍뚝썰기/300g)   1.0  4977  매장A  10:41  NaN  
2  유정란/매장용(10알/국내산)   1.0  7083  매장A  11:00  NaN  
3     도토리묵(420g:1모)   1.0  7083  매장A  11:28  NaN  
4  유정란/매장용(10알/국내산)   1.0  7083  매장A  11:59  NaN

# 3. 데이터 전처리
- 각 데이터의 결측치/이상치 확인 및 발생원인 파악
- 결측치/이상치 처리방법 검토 및 적용

### (1) 고객 정보 데이터 결측치 확인 및 처리

In [ ]:
# 고객 정보 데이터 결측값 확인
df.isnull().sum()

회원번호           0
조합원상태          0
주소             0
성별             2
연령             2
배송서비스신청여부      0
모바일알람여부        0
Gold_member    0
dtype: int64

총 11555명의 회원 데이터 중에서 성별에서 2명, 연령에서 2명 결측치 확인

In [ ]:
# 연령 칼럼에서 결측값이 있는 행들을 출력
missing_age = df[df['연령'].isna()]
missing_age

회원번호 조합원상태     주소   성별  연령 배송서비스신청여부 모바일알람여부 Gold_member
5983  383599936  정상회원  기타-기타    여 NaN       미신청       .      normal
7098  545864448  정상회원  기타-기타  NaN NaN       미신청       .      normal

In [ ]:
# 성별 칼럼에서 결측값이 있는 행들을 출력
missing_gender = df[df['성별'].isna()]
missing_gender

회원번호 조합원상태     주소   성별    연령 배송서비스신청여부 모바일알람여부 Gold_member
3666  1527053696  정상회원  기타-기타  NaN  45.0       미신청       .      normal
7098   545864448  정상회원  기타-기타  NaN   NaN       미신청       .      normal

In [ ]:
# 성별과 연령대의 결측값 제거
df = df.drop(index=3666)
df = df.drop(index=7098)
df = df.drop(index=5983)

# 고객 데이터 결측치 재확인
df.isnull().sum()

회원번호           0
조합원상태          0
주소             0
성별             0
연령             0
배송서비스신청여부      0
모바일알람여부        0
Gold_member    0
dtype: int64

성별, 연령대 결측값 제거. 7098 회원은 성별, 나이 둘다 없는 고객이므로 총 3개 행 제거.

### (2) 판매 정보 데이터 결측치 확인 및 처리

In [ ]:
# 판매 정보 데이터 결측값 확인
dfm.isnull().sum()

Unnamed: 0.1         0
Unnamed: 0           0
공급일자                 0
회원번호                 0
조합원상태                0
물품대분류                0
물품중분류                0
물품소분류                0
물품명                  0
구매수량                 0
구매금액                 0
구매매장                 0
구매시각                 0
반품일자            289685
dtype: int64

반품일자에 289685개의 결측치가 있음을 확인하였다. 반품일자가 없는 행(NaN)은 반품을 하지 않은 경우이기 때문에 결측치를 처리할 필요가 없다.

### (3) 고객 정보 데이터 이상치 확인 및 처리
- 조합원상태 이상치 데이터
- 구매금액이 음수인 데이터
- 주소 이상치 데이터
- 판매 정보 데이터의 이상치 데이터

#### 조합원상태 데이터

In [ ]:
# 판매 정보 데이터 조합원상태 확인
dfm["조합원상태"].value_counts() #조합원상태가 탈퇴, 탈퇴처리중, 탈퇴신청으로 나뉜다.

정상회원     291371
탈퇴         3216
탈퇴처리중        53
탈퇴신청          5
Name: 조합원상태, dtype: int64

In [ ]:
# 고객 정보 데이터 조합원상태 확인
df["조합원상태"].value_counts()

정상회원     11410
탈퇴         139
탈퇴처리중        2
탈퇴신청         1
Name: 조합원상태, dtype: int64

In [ ]:
# 조합원 상태중 탈퇴처리중 탈퇴신청의경우 탈퇴한 회원으로 처리
dfm['조합원상태'] = dfm['조합원상태'].apply(lambda x: x if x == '정상회원' else '탈퇴회원')
df['조합원상태'] = df['조합원상태'].apply(lambda x: x if x == '정상회원' else '탈퇴회원')

In [ ]:
# 판매 정보 데이터 조합원상태 재확인
dfm["조합원상태"].value_counts()

정상회원    291371
탈퇴회원      3274
Name: 조합원상태, dtype: int64

In [ ]:
# 고객 정보 데이터 조합원상태 재확인
df["조합원상태"].value_counts()

정상회원    11410
탈퇴회원      142
Name: 조합원상태, dtype: int64

#### 구매금액 데이터

In [ ]:
# 구매금액이 음수인 행 확인
negative_purchases = dfm[dfm['구매금액'] < 0]

# 해당 행 제거
dfm = dfm.drop(negative_purchases.index)

#### 주소 데이터

In [ ]:
# 주소를 구/동 별로 범주화
df[['구', '동']] = df['주소'].str.split('-', 1, expand=True)

In [ ]:
# 주소 이상치 처리
df.loc[df['주소'] == '기타-금광동', '구'] = '중원구'
df.loc[df['주소'] == '기타-신흥동', '구'] = '수정구'
df.loc[df['주소'] == '기타-금곡동', '구'] = '분당구'
df.loc[df['주소'] == '기타-동천동', '구'] = '수지구'
df.loc[df['주소'] == '기타-중앙동', '구'] = '중원구'
df.loc[df['주소'] == '기타-보라동', '구'] = '기흥구'
df.loc[df['주소'] == '기타-동백동', '구'] = '기흥구'
df.loc[df['주소'] == '기타-은행동', '구'] = '중원구'
df.loc[df['주소'] == '기타-궁내동', '구'] = '분당구'
df.loc[df['주소'] == '기타-율동', '구'] = '분당구'
df.loc[df['주소'] == '기타-태평동', '구'] = '수정구'
df.loc[df['주소'] == '기타-성남동', '구'] = '기흥구'
df.loc[df['주소'] == '기타-비조합원', '동'] = '기타'

### (4) 판매 정보 데이터 이상치 확인 및 처리

In [ ]:
# merge 과정에서 생긴 의미 없는 열 제거
dfm.drop(['Unnamed: 0.1', 'Unnamed: 0'], axis=1, inplace=True)

In [ ]:
# 물품명이 올바르지 않은 행 제거
dfm[dfm["물품명"] == "조합원의식조사설문"]
dfm = dfm[dfm["물품명"] != "조합원의식조사설문"]

In [ ]:
# 물품명이 올바르지 않으며 가격 예측이 불가하여 제거
dfm[(dfm["물품명"] == "통삼겹-성남용인") & (dfm["구매금액"] == 0)]
dfm = dfm.drop(index=134334)

In [ ]:
#반품일자가 있으나 반품이 아닌 구매 수량이 양수인 4개의 행 제거
dfm[(dfm["반품일자"].notna()) & (dfm['구매수량'] > 0)]
dfm = dfm.drop(index = 849)
dfm = dfm.drop(index = 105556)
dfm = dfm.drop(index = 161974)
dfm = dfm.drop(index = 261634)

In [ ]:
# 구매수량은 있으나 구매 금액이 0인 증정용 데이터 확인
dfm[(dfm['구매금액'] == 0) & (dfm['구매수량'] == 1)]

# 데이터가 몰려있긴하지만 매출에 크게 변화가 없기 때문에 제거
item_names = dfm[(dfm['구매금액'] == 0) & (dfm['구매수량'] == 1)]['물품명']

In [ ]:
# 구매 수량이 음수인 데이터 확인 및 제거
dfm[(dfm["구매수량"] < 0)]
dfm = dfm[dfm['구매금액'] != 0]

### (5) 파생변수 생성
- 판매 정보 데이터 : 유기농 물품 변수 생성
- 고객 정보 데이터 : 연령대 변수, 지역 변수 생성

#### 유기농 물품 변수 생성

In [ ]:
# 유기농 제품의 정보를 확인하고 분류
pattern = r'/유기|유기농|유|논지엠오|무농약|무항생제'
dfm['유기농물품'] = dfm['물품명'].str.contains(pattern, regex=True)

# '유기농 물품' 열을 범주형 'O'와 'X'로 변환
dfm['유기농 물품'] = dfm['유기농물품'].apply(lambda x: 'O' if x else 'X')
del dfm['유기농물품']

# 수정된 데이터프레임을 출력
dfm.head()

공급일자       회원번호 조합원상태 물품대분류  물품중분류   물품소분류               물품명  구매수량  \
0  2023-01-06  220954944  정상회원    반찬  두부/유부      두부       두부(1모:420g)   1.0   
1  2023-01-06   89023040  정상회원    채소   열매채소  단호박/손질    단호박(깍뚝썰기/300g)   1.0   
2  2023-01-06  354941440  정상회원   축산물      알     유정란  유정란/매장용(10알/국내산)   1.0   
3  2023-01-06  786634048  탈퇴회원    반찬   어묵/묵       묵     도토리묵(420g:1모)   1.0   
4  2023-01-06  438543104  정상회원   축산물      알     유정란  유정란/매장용(10알/국내산)   1.0   

   구매금액 구매매장   구매시각 반품일자 유기농 물품  
0  3829  매장A  10:32  NaN      X  
1  4977  매장A  10:41  NaN      X  
2  7083  매장A  11:00  NaN      O  
3  7083  매장A  11:28  NaN      X  
4  7083  매장A  11:59  NaN      O

#### 연령대 변수 생성

In [ ]:
# 연령을 10년 단위로 나이대 분류하여 연령대 변수 생성
bins = [0, 29, 39, 49, 59, 69, float('inf')]
labels = ['20대 이하', '30대', '40대', '50대', '60대', '70대 이상']
df['연령대'] = pd.cut(df['연령'], bins=bins, labels=labels, right=False)
df.head()

회원번호 조합원상태        주소 성별    연령 배송서비스신청여부 모바일알람여부 Gold_member    구  \
0  1000012800  정상회원  수지구-풍덕천동  여  58.0       미신청       .      normal  수지구   
1  1000103744  정상회원   처인구-고림동  여  34.0       미신청       .      normal  처인구   
2  1000122304  정상회원     광주-기타  여  38.0       미신청       .      normal   광주   
3  1000163136  정상회원   수지구-동천동  여  73.0       미신청       .      normal  수지구   
4  1000281920  정상회원  수지구-풍덕천동  여  42.0        신청       .      normal  수지구   

      동     연령대  
0  풍덕천동     50대  
1   고림동     30대  
2    기타     30대  
3   동천동  70대 이상  
4  풍덕천동     40대

In [ ]:
# # 연령을 더욱 자세한 범주로 분류
# bins = [0, 25, 30, 35, 40, 45,50,55,60,65,70,float('inf')]
# labels = ['20대초반','20대후반' ,'30대초반','30대후반','40대초반','40대후반','50대초반','50대후반', '60대초반', '60대후반', '70대']
# df['연령대_세분화'] = pd.cut(df['연령'], bins=bins, labels=labels, right=False)

#### 지역 변수 생성

In [ ]:
# 구 분류 기준 정의
# (1) '수지'를 제외한 '용인시'와 인접한 '수원시', '성남시'의 지역을 주변지역으로 분류
# (2) 그 외의 지역들을 '먼 지역'으로 분류

def classify_region(gu):
    suji_region = ['수지구']
    nearby_region = ['기흥구', '분당구', '수원 영통구', '수원 팔달구', '수원 장안구', '수원 권선구',  '처인구','중원구', '수정구']
    distant_region = ['서울 강남구', '서울 송파구', '화성', '광주', '하남','기타']

    if gu in suji_region:
        return '수지 지역'
    elif gu in nearby_region:
        return '주변 지역'
    elif gu in distant_region:
        return '먼 지역'
    else:
        return '기타'

# 지역 칼럼 추가
df['지역'] = df['구'].apply(classify_region)

# 4. 전처리 후 최종 데이터 정보

### (1) 고객 정보 데이터
- 고객 정보 데이터(Row : 11,552 / Column : 12) | 변수명 : df

| 변수명             | 타입   | 상세내용                                                 |
| ------------------ | ------ | -------------------------------------------------------- |
| 회원번호           | ID     | 회원 고유 식별 번호                                      |
| 조합원상태         | 범주형 | 회원 가입 상태(탈퇴여부)                                 |
| 주소               | 범주형 | 회원 주소지                                              |
| 성별               | 범주형 | 회원성별                                                 |
| 연령               | 연속형 | 회원 연령                                                |
| 연령대             | 범주형 | 고객 나이 연령대                                         |
| 배송서비스신청여부 | 범주형 | 매장에서 물품 구매 시 집까지 배송서비스 신청 여부        |
| 모바일알람신청여부 | 범주형 | 프로모션 및 할인 관련, 마케팅 관련 메시지 알람 수신 여부 |
| Gold_member        | 범주형 | VIP 고객 여부                                            |
| 구                 | 범주형 | 고객 주소지의 구                                            |
| 동                 | 범주형 | 고객 주소지의 동                                            |
| 지역               | 범주형 | 고객 주소지의 지역                                            |


### (2) 판매 정보 데이터
- 판매 정보 데이터(Row : 288,814 / Column : 13) | 변수명 : dfm

| 변수명           | 타입     | 상세내용                                    |
| ---------------- | -------- | ------------------------------------------- |
| 공급일자         | datetime | 상품 판매 일자                              |
| 회원번호         | ID       | 회원 고유 식별 ID                           |
| 조합원상태       | 범주형   | 회원 가입 상태(탈퇴여부)                    |
| 물품대분류       | 범주형   | 구매 제품의 물품 대분류                     |
| 물품중분류       | 범주형   | 구매 제품의 물품 중분류                     |
| 물품소분류       | 범주형   | 구매 제품의 물품 소분류                     |
| 물품명           | 범주형   | 구매 제품의 물품명                          |
| 구매수량         | 연속형   | 고객이 구매한 물품 수량                     |
| 구매금액         | 연속형   | 고객이 구매한 물품 금액( 원단위)            |
| 구매매장         | 범주형   | 구매 매장                                   |
| 반품일자 | datetime | 고객이 품목을 반품한 경우, 반품 날짜 (환불) |
| 구매시각         | time     | 구매시각 (바코드에 상품이 찍힌 시간)        |
| 유기농 물품       | 범주형     | 유기농 제품 여부(O, X)             |

# 5. 전처리 완료된 파일 생성

In [ ]:
df.to_csv('pre_member_data', index=False)
dfm.to_csv('pre_data.csv', index=False)